### PR TITLE
[Request for contributions] implement get_reachable_from_inputs

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
### Summary

Hello, this PR implements the function `get_reachable_from_inputs` as described in this ticket: https://github.com/keras-team/keras/projects/1#card-9317506

### Related Issues

https://github.com/keras-team/keras/projects/1#card-9317506
#11777

### PR Overview

* Tests were added

* Tensorflow backend is working (implementation taken from tf.keras)

* Theano backend is working if there are target tensors. This is because Theano tensors have only pointers to previous tensors, not the next ones. So we can't do the same as in the tensorflow implementation. We have to start from the outputs, come back (like the backpropagation), add pointers, and then we can do the same algorithm as in the tensorflow backend.

* I didn't manage to get the CNTK backend to work as I didn't find a way to get children or parents nodes from CNTK functions and variables. @souptc, help would be very much welcome. Thank you in advance. 

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [x] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by souptc** [comment]
I am not sure what is the recommended way of get parent/child node through cntk python API, add @liqunfu, who is from cntk team and more familiar with the API.

**Comment by gabrieldemarmiesse** [comment]
Thank you for the quick response!

File: pytest.ini [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Configuration of py.test
 [pytest]
 addopts=-v
-        -n 2
+        -n 1
```
Comment (Dref360) [inline pytest.ini]: set back to 2

File: keras/backend/theano_backend.py [hunk 2]
```diff
@@ -2840,3 +2841,77 @@ def local_conv2d(inputs, kernel, kernel_size, strides, output_shape, data_format
...
+
+
+def get_reachable_from_inputs(inputs, targets=None):
+    if targets is None:
+        raise NotImplementedError('It is not possible to use '
+                                  '`get_reachable_from_inputs` with the Theano'
+                                  ' backend without specifying target tensors to'
+                                  ' reach.')
+
+    backward_pass(targets)
```
Comment (Dref360) [inline keras/backend/theano_backend.py]: Is there a way to do it in a one-step process?

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -2840,3 +2841,77 @@ def local_conv2d(inputs, kernel, kernel_size, strides, output_shape, data_format
...
+    return get_reachable_from_inputs_with_ptrs(inputs, targets)
+
+
+def backward_pass(outputs):
+    queue = outputs[:]
```
Comment (Dref360) [inline keras/backend/theano_backend.py]: What's the point of [:]?

File: keras/backend/theano_backend.py [hunk 4]
```diff
@@ -2840,3 +2841,77 @@ def local_conv2d(inputs, kernel, kernel_size, strides, output_shape, data_format
...
+        inputs = link_to_inputs(x)
+        for input_ in inputs:
+            queue.insert(0, input_)
+
+
+def link_to_inputs(node):
+    try:
+        assert node._keras_done
+        return []
+    except AttributeError:
```
Comment (Dref360) [inline keras/backend/theano_backend.py]: We should use hasattr

File: pytest.ini (hunk 1 above)
Comment (gabrieldemarmiesse) [reply pytest.ini]: oups, my bad.

File: keras/backend/theano_backend.py (hunk 3 above)
Comment (gabrieldemarmiesse) [reply keras/backend/theano_backend.py]: to copy the input list. We'll modify it later, and we don't want to modify the argument directly.

**Comment by gabrieldemarmiesse** [comment]
Since it seems that it's not currently possible, I'll close this PR.

**Pull request closed without merging**